### PR TITLE
AP_Param: move complex AP_ParamT functions to cpp

### PR DESCRIPTION
**Pull request by IamPete1**
Moves param template functions into a new CPP file to save flash.

### CODE REVIEW

File: libraries/AP_Param/AP_Param.h [hunk 1]
```diff
@@ -895,9 +855,7 @@ class AP_ParamT : public AP_Param
 
     /// AP_ParamT types can implement AP_Param::cast_to_float
     ///
-    float cast_to_float(void) const {
-        return (float)_value;
-    }
+    float cast_to_float(void) const;
```
Comment (andyp1per) [inline libraries/AP_Param/AP_Param.h:858]: this should stay as an inline IMO, I don't think this will save any flash

Comment (IamPete1) [reply libraries/AP_Param/AP_Param.h:858]: Interestingly, this move saves 320 bytes on CubeOrange but costs on 40 on periph. I guess its better to save a little on the small boards than a lot on the big boards. I will wait for the CI size test to complete before pushing as we have a better comparison.

Comment (IamPete1) [reply libraries/AP_Param/AP_Param.h:858]: As it is now:

Periph:
```
Binary Name    Text [B]        Data [B]     BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
-------------  --------------  -----------  -------------  ----------------------------  -------------------------
ap_periph      220 (+0.2190%)  0 (0.0000%)  -4 (-0.0199%)  220 (+0.2186%)                                     4616
```
Pixhawk 1- 1M
```
Binary Name      Text [B]          Data [B]     BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  ----------------  -----------  -------------  ----------------------------  -------------------------
antennatracker   -540 (-0.0720%)   0 (0.0000%)  4 (+0.0020%)   -540 (-0.0719%)                                  281816
arducopter-heli  -1096 (-0.1084%)  0 (0.0000%)  0 (0.0000%)    -1096 (-0.1083%)                                  21124
ardurover        -1168 (-0.1274%)  0 (0.0000%)  0 (0.0000%)    -1168 (-0.1273%)                                 115556
arducopter       -1268 (-0.1269%)  0 (0.0000%)  -4 (-0.0020%)  -1268 (-0.1268%)                                  33532
ardusub          -1160 (-0.1263%)  0 (0.0000%)  0 (0.0000%)    -1160 (-0.1262%)                                 114040
blimp            -444 (-0.0627%)   0 (0.0000%)  -4 (-0.0020%)  -444 (-0.0627%)                                  324124
arduplane        -1304 (-0.1267%)  0 (0.0000%)  0 (0.0000%)    -1304 (-0.1266%)                                   3284
```
F405
```
Binary Name      Text [B]          Data [B]     BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  ----------------  -----------  -------------  ----------------------------  -------------------------
antennatracker   -620 (-0.0945%)   0 (0.0000%)  4 (+0.0031%)   -620 (-0.0944%)                                  326796
arducopter-heli  -1312 (-0.1388%)  0 (0.0000%)  0 (0.0000%)    -1312 (-0.1385%)                                  36972
ardurover        -1376 (-0.1613%)  0 (0.0000%)  0 (0.0000%)    -1376 (-0.1612%)                                 130692
arducopter       -1484 (-0.1587%)  0 (0.0000%)  -4 (-0.0031%)  -1484 (-0.1583%)                                  47244
ardusub          -1348 (-0.1656%)  0 (0.0000%)  -4 (-0.0031%)  -1348 (-0.1655%)                                 169736
blimp            -596 (-0.0883%)   0 (0.0000%)  4 (+0.0031%)   -596 (-0.0882%)                                  307552
arduplane        -1556 (-0.1596%)  0 (0.0000%)  4 (+0.0031%)   -1556 (-0.1593%)                                   7516
```
Durandal
```
Binary Name      Text [B]          Data [B]     BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  ----------------  -----------  -------------  ----------------------------  -------------------------
antennatracker   -3256 (-0.2406%)  0 (0.0000%)  0 (0.0000%)    -3256 (-0.2402%)                                 613972
arducopter-heli  -4848 (-0.2714%)  0 (0.0000%)  0 (0.0000%)    -4848 (-0.2709%)                                 181548
ardurover        -4360 (-0.2719%)  0 (0.0000%)  0 (0.0000%)    -4360 (-0.2716%)                                 364876
arducopter       -5236 (-0.2944%)  0 (0.0000%)  4 (+0.0015%)   -5236 (-0.2938%)                                 189372
ardusub          -4596 (-0.2937%)  0 (0.0000%)  -4 (-0.0015%)  -4596 (-0.2934%)                                 404104
blimp            -3240 (-0.2500%)  0 (0.0000%)  0 (0.0000%)    -3240 (-0.2497%)                                 671712
arduplane        -5172 (-0.2936%)  0 (0.0000%)  4 (+0.0015%)   -5172 (-0.29%)                                 206708
```

Comment (IamPete1) [reply libraries/AP_Param/AP_Param.h]: So, adding back this method has saved 32 bytes on periph, 40 bytes on Pixhawk1 and F405 but cost us 300 bytes on Durandal.

**Comment by IamPete1** [comment]
Rebased to see if this will fit on periph now.

**Comment by IamPete1** [comment]
Still too big for periph, I guess the question is what is more important, a little space on periph or lots of space on other boards. I'm sure we could find the flash to get this to fit on periph if we decided it was worth it.

**Comment by andyp1per** [comment]
@IamPete1 I suspect you could do something similar to what I did for vector ops to make this fit on periph:
https://github.com/ArduPilot/ardupilot/blob/427c64eff0a2d05c6fed80026d55c02bc4616c52/libraries/AP_Math/vector3.h#L327

**Comment by IamPete1** [comment]
Durandal:
```
Binary Name      Text [B]          Data [B]     BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  ----------------  -----------  -------------  ----------------------------  -------------------------
blimp            -3240 (-0.2495%)  0 (0.0000%)  0 (0.0000%)    -3240 (-0.2492%)                                 669072
ardusub          -4480 (-0.2856%)  0 (0.0000%)  0 (0.0000%)    -4480 (-0.2852%)                                 399928
ardurover        -4244 (-0.2626%)  0 (0.0000%)  -4 (-0.0015%)  -4244 (-0.2623%)                                 352068
antennatracker   -3260 (-0.2404%)  0 (0.0000%)  4 (+0.0015%)   -3260 (-0.2400%)                                 611228
arduplane        -5052 (-0.2862%)  0 (0.0000%)  4 (+0.0015%)   -5052 (-0.2856%)                                 202420
arducopter-heli  -4720 (-0.2635%)  0 (0.0000%)  0 (0.0000%)    -4720 (-0.2630%)                                 176300
arducopter       -5100 (-0.2858%)  0 (0.0000%)  4 (+0.0015%)   -5100 (-0.2852%)                                 183244
```
MatekF405:
```
Binary Name      Text [B]          Data [B]     BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  ----------------  -----------  -------------  ----------------------------  -------------------------
blimp            -636 (-0.0938%)   0 (0.0000%)  4 (+0.0031%)   -636 (-0.0937%)                                  304568
ardusub          -1440 (-0.1761%)  0 (0.0000%)  0 (0.0000%)    -1440 (-0.1759%)                                 165984
ardurover        -1460 (-0.1704%)  0 (0.0000%)  4 (+0.0031%)   -1460 (-0.1702%)                                 126676
antennatracker   -660 (-0.1001%)   0 (0.0000%)  -4 (-0.0031%)  -660 (-0.1000%)                                  323868
arduplane        -1640 (-0.1681%)  0 (0.0000%)  0 (0.0000%)    -1640 (-0.1677%)                                   6724
arducopter-heli  -1400 (-0.1474%)  0 (0.0000%)  0 (0.0000%)    -1400 (-0.1470%)                                  32060
arducopter       -1576 (-0.1675%)  0 (0.0000%)  0 (0.0000%)    -1576 (-0.1671%)                                  41564
```
Pixhawk1-1M
```
Binary Name      Text [B]          Data [B]     BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  ----------------  -----------  -------------  ----------------------------  -------------------------
blimp            -480 (-0.0678%)   0 (0.0000%)  0 (0.0000%)    -480 (-0.0677%)                                  323660
ardusub          -1204 (-0.1309%)  0 (0.0000%)  -4 (-0.0020%)  -1204 (-0.1308%)                                 112640
ardurover        -1204 (-0.1299%)  0 (0.0000%)  -4 (-0.0020%)  -1204 (-0.1298%)                                 105556
antennatracker   -576 (-0.0767%)   0 (0.0000%)  0 (0.0000%)    -576 (-0.0767%)                                  281336
arduplane        -1340 (-0.1300%)  0 (0.0000%)  4 (+0.0020%)   -1340 (-0.1299%)                                   1756
arducopter-heli  -1132 (-0.1117%)  0 (0.0000%)  -4 (-0.0020%)  -1132 (-0.1116%)                                  18652
arducopter       -1304 (-0.1301%)  0 (0.0000%)  0 (0.0000%)    -1304 (-0.1300%)                                  30276
```
Hitec-Airspeed:
```
Binary Name    Text [B]        Data [B]     BSS (B)       Total Flash Change [B] (%)      Flash Free After PR (B)
-------------  --------------  -----------  ------------  ----------------------------  -------------------------
ap_periph      308 (+0.3267%)  0 (0.0000%)  4 (+0.0184%)  308 (+0.3248%)                                     9320
```
f103-GPS
```
Binary Name    Text [B]        Data [B]     BSS (B)      Total Flash Change [B] (%)      Flash Free After PR (B)
-------------  --------------  -----------  -----------  ----------------------------  -------------------------
ap_periph      224 (+0.2221%)  0 (0.0000%)  0 (0.0000%)  224 (+0.2218%)                                     4248
```

**Review by tridge — APPROVED** [review]

**Pull request merged**